diff --git a/dist/esm/request-response.js b/dist/esm/request-response.js
index 4ce23fae2de3772646a2380a9e64c3f468b051f6..5e147cb350eacc4e35ffa93971dd1bc064cf1cbd 100644
--- a/dist/esm/request-response.js
+++ b/dist/esm/request-response.js
@@ -31,6 +31,13 @@ function attachResponseHeaders(value, event) {
 	if (value instanceof Response) mergeEventResponseHeaders(value, event);
 	return value;
 }
+function handleRequestAbort(value, request) {
+	if (!isPromiseLike(value)) return value;
+	return Promise.resolve(value).catch((error) => {
+		if (request.signal.aborted && error === request.signal.reason) return new Response(null, { status: 499 });
+		throw error;
+	});
+}
 function requestHandler(handler) {
 	return (request, requestOpts) => {
 		let h3Event;
@@ -43,8 +50,9 @@ function requestHandler(handler) {
 			});
 			throw error;
 		}
-		return toResponse(attachResponseHeaders(eventStorage.run({ h3Event }, () => handler(request, requestOpts)), h3Event), h3Event);
+		const response = handleRequestAbort(eventStorage.run({ h3Event }, () => handler(request, requestOpts)), request);
+		return toResponse(attachResponseHeaders(response, h3Event), h3Event);
 	};
 }
 function getH3Event() {
 	const event = eventStorage.getStore();
diff --git a/src/request-response.ts b/src/request-response.ts
index e0ad8a336a37bc29ccdff48d87dfb5b360b633f1..0183d0b4ff5fe92d6215851f96567da743ecd50d 100644
--- a/src/request-response.ts
+++ b/src/request-response.ts
@@ -118,6 +118,22 @@ function attachResponseHeaders<T>(
   return value
 }
 
+function handleRequestAbort<T>(
+  value: MaybePromise<T>,
+  request: Request,
+): MaybePromise<T | Response> {
+  if (!isPromiseLike(value)) {
+    return value
+  }
+
+  return Promise.resolve(value).catch((error) => {
+    if (request.signal.aborted && error === request.signal.reason) {
+      return new Response(null, { status: 499 })
+    }
+    throw error
+  })
+}
+
 export function requestHandler<TRegister = unknown>(
   handler: RequestHandler<TRegister>,
 ) {
@@ -135,9 +151,12 @@ export function requestHandler<TRegister = unknown>(
       throw error
     }
 
-    const response = eventStorage.run({ h3Event }, () =>
-      handler(request, requestOpts),
+    const response = handleRequestAbort(
+      eventStorage.run({ h3Event }, () =>
+        handler(request, requestOpts),
+      ),
+      request,
     )
     return h3_toResponse(attachResponseHeaders(response, h3Event), h3Event)
   }
 }
